presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -389,13 +388,47 @@ public synchronized void finishInsertIntoExistingTable(
                 throw new TableNotFoundException(schemaTableName);
             case ADD:
             case ALTER:
-            case INSERT_EXISTING:
+            case INSERT_EXISTING: {
+                // Handle create partitioned table + insert in a single transaction
+                if (oldTableAction.getData().getTable().isPartitioned()) {
+                    // When inserting into partitioned table no files should be added at the table level
+                    verify(isEmpty(oldTableAction.getData().getFileNames()), "partitioned table got files to move on a table level");
+                    verify(isEmpty(fileNames), "unexpected files to move on a table level for partitioned table");
+                    updateTableAction(oldTableAction, (table) -> incrementStatisticsCounters(table, statistics));
+                    return;
+                }
                 throw new UnsupportedOperationException("Inserting into an unpartitioned table that were added, altered, or inserted into in the same transaction is not supported");
+            }
             default:
                 throw new IllegalStateException("Unknown action type");
         }
     }
 
+    private void updateTableAction(Action<TableAndMore> action, Function<Table, Table> update)
+    {
+        TableAndMore tableAndMore = action.getData();
+        Table table = tableAndMore.getTable();
+        Table updatedTable = update.apply(table);
+        verify(table.getDatabaseName().equals(updatedTable.getDatabaseName()), "changing database name is not allowed");
+        verify(table.getTableName().equals(updatedTable.getTableName()), "changing table name is not allowed");
+        TableAndMore updatedTableAndMore = new TableAndMore(
+                updatedTable,
+                tableAndMore.principalPrivileges,
+                tableAndMore.getCurrentLocation(),
+                tableAndMore.getFileNames(),
+                tableAndMore.isIgnoreExisting());
+        tableActions.put(
+                new SchemaTableName(table.getDatabaseName(), table.getTableName()),
+                new Action<>(
+                        action.getType(),
+                        updatedTableAndMore, action.getContext()));
+    }
+
+    private static <T> boolean isEmpty(Optional<List<T>> list)
+    {
+        return !list.isPresent() || list.get().isEmpty();
```

Comment:
`list.map(List::isEmpty).or(true)`?

Reply:
I don't think it is more readable. But i don't have any strong opinion.